File: langs/csharp/src/VMx/Commands/CompositeCommand.cs
using System.Windows.Input;

namespace VMx.Commands;

/// <summary>
/// Aggregates N inner <see cref="ICommand"/> instances. See spec/04-commands.md
/// §Decorators and ADR-0012.
///
/// <list type="bullet">
/// <item><c>CanExecute</c> returns true iff at least one inner returns true.</item>
/// <item><c>Execute</c> invokes every inner whose <c>CanExecute</c> is true.</item>
/// <item><c>CanExecuteChanged</c> fires when any inner's CanExecuteChanged fires.</item>
/// </list>
/// </summary>
public sealed class CompositeCommand : ICommand, IDisposable
{
    private readonly IReadOnlyList<ICommand> _inner;
    private readonly List<EventHandler> _innerHandlers = new();
    private bool _disposed;

    /// <summary>Creates a new <see cref="CompositeCommand"/> over the given inner commands.</summary>
    public CompositeCommand(params ICommand[] inner)
        : this((IReadOnlyList<ICommand>)inner)
    {
    }

    /// <summary>Creates a new <see cref="CompositeCommand"/> over the given inner commands.</summary>
    public CompositeCommand(IReadOnlyList<ICommand> inner)
    {
        _inner = inner ?? throw new ArgumentNullException(nameof(inner));
        foreach (var c in _inner)
        {
            EventHandler handler = (sender, args) => CanExecuteChanged?.Invoke(this, args);
            _innerHandlers.Add(handler);
            c.CanExecuteChanged += handler;
        }
    }

    /// <inheritdoc/>
    public event EventHandler? CanExecuteChanged;

    /// <inheritdoc/>
    public bool CanExecute(object? parameter)
    {
        foreach (var c in _inner)
            if (c.CanExecute(parameter))
                return true;
        return false;
    }

    /// <inheritdoc/>
    public void Execute(object? parameter)
    {
        foreach (var c in _inner)
            if (c.CanExecute(parameter))
                c.Execute(parameter);
    }

    /// <summary>Unsubscribes from inner <c>CanExecuteChanged</c> events.</summary>
    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        for (var i = 0; i < _inner.Count; i++)
            _inner[i].CanExecuteChanged -= _innerHandlers[i];
    }
}

File: langs/python/src/vmx/commands/composite_command.py
"""CompositeCommand — aggregates N inner commands.

See spec/04-commands.md §Decorators and ADR-0012.
"""

from __future__ import annotations

from collections.abc import Sequence
from typing import Any

import reactivex as rx
from reactivex import Observable
from reactivex.abc import DisposableBase

from vmx.commands.protocols import Command


class CompositeCommand:
    """Aggregates N inner commands.

    - ``can_execute()`` returns True iff at least one inner returns True.
    - ``execute()`` invokes every inner whose ``can_execute`` is True.
    - ``can_execute_changed`` fires when any inner's stream fires.
    """

    def __init__(self, *inner: Command) -> None:
        self._inner: Sequence[Command] = inner
        self._subscriptions: list[DisposableBase] = []
        self._disposed = False
        self._can_execute_changed: Observable[None]
        if not inner:
            self._can_execute_changed = rx.never()
        else:
            self._can_execute_changed = rx.merge(*(c.can_execute_changed for c in inner))

    @property
    def can_execute_changed(self) -> Observable[None]:
        return self._can_execute_changed

    def can_execute(self, parameter: Any = None) -> bool:
        return any(c.can_execute(parameter) for c in self._inner)

    def execute(self, parameter: Any = None) -> None:
        for c in self._inner:
            if c.can_execute(parameter):
                c.execute(parameter)

    def dispose(self) -> None:
        """Dispose internal subscriptions. Idempotent."""
        if self._disposed:
            return
        self._disposed = True
        for s in self._subscriptions:
            s.dispose()
        self._subscriptions.clear()
